Add article about Retroactive default StorageClass assignement (#35658)
kubernetes/website: content/en/docs/concepts/storage/persistent-volumes.md

the API server component. For more information on API server command-line flags,
check [kube-apiserver](/docs/admin/kube-apiserver/) documentation.

> ### Retroactive StorageClass assignment

Reviewer: I'd like to move this much later down the page - perhaps just before “Volume populators and data sources”.

Someone who is trying to learn about PVs and PVCs work is going to find this detail is mentioned too early in their learning journey.

Author: Done.

the API server component. For more information on API server command-line flags,
check [kube-apiserver](/docs/admin/kube-apiserver/) documentation.

> Claims can also be created without any StorageClass while no default exists. In that case
> their StorageClass will remain unset until a default becomes available, then the claim
> will get the new default assigned by `RetroactiveDefaultStorageClass` [feature](/docs/concepts/storage/persistent-volumes/#retroactive-storageclass-assignment).
> 
> ### Retroactive StorageClass assignment

Reviewer: ```suggestion
### Retroactive StorageClass assignment

You can create a PersistentVolumeClaim without specifying a `storageClassName` for the new
PVC, and you can do so even when no default StorageClass exists in your cluster.
In this case, the new PVC creates as you defined it, and the `storageClassName` of that PVC
remains set to `""` until a default becomes available. If or when you or somebody else marks a
StorageClass as default for your cluster, that PVC continues to have the `storageClassName: ""`.
However, if you enable the `RetroactiveDefaultStorageClass` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) then Kubernetes behaves differently: existing PVCs that have an
empty `storageClassName` update to use the new default StorageClass.
```

Author: It seems that this file combines both 1st and 3rd person forms, however the surrounding articles are in the 3rd person form so I tried to maintain that. Maybe there is some rule related to this?

Also I think the first form can be confusing, because sometimes articles describe operations that are typically done by administrator of the cluster (with elevated priviliges) and operations done by a regular user. For example in case of classes and volumes, a class would be created by admin and PVC by user.

Lastly the sentence suggested:

"You can create a PersistentVolumeClaim without specifying a `storageClassName`...

...In this case, the new PVC creates as you defined it, and the `storageClassName` of that PVC remains set to `""` until a default becomes available."

It's not very accurate, you don't magically get `""` when you don't specify class name in PVC. Because "without specifying" can mean either leaving out `storageClassName` key out or setting it to nil value or `""`. All these are allowed when creating PVCs and it's important to distinguish those because historically this was the same thing but not anymore. Only the case with `""` truly means that user wants no class. This is described somewhere around L74.

Reviewer: > it's important to distinguish those because historically this was the same thing but not anymore

Let's call that out. It wasn't obvious to me.

Author: Ack, moved under "Retroactive StorageClass assignment" and changed to active form.

{{< feature-state for_k8s_version="v1.25" state="alpha" >}}

A PVC can be created without specifying any StorageClass. This can happen either if there is no default StorageClass set by the administrator or [`DefaultStorageClass` admission plugin](/docs/reference/access-authn-authz/admission-controllers/#defaultstorageclass) is disabled. PVs with `StorageClassName` set to `""` will continue to bind to such PVCs until default StorageClass is created.

> When default StorageClass becomes available any existing PVCs without `StorageClassName` will get this value updated retroactively to the default StorageClass name. In order to keep binding to PVs with `StorageClassName` set to `""` (while default StorageClass is present) users need to set PVC `StorageClassName` to `""`.

Reviewer: When a default StorageClass becomes available, the control plane identifies any existing PVCs with an empty`storageClassName`. For the PVCs that have an empty value `storageClassName`, the control plane then updats those PVCs to set `sotrageClassName` to match the new default StorageClass.

In order to keep binding to PVs with `storageClassName` set to `""` (while default StorageClass is present), you need to set the `storageClassName` of the associated PVC to `""`.
```

Author: Same as above, the word empty needs to be used with caution. Because empty string (`""`) is exactly and the only one case when PVC will **not** get updated.

Author: Looks good, I've updated it to match the form suggested and added more details.

> This behavior helps administrators change default storage class by removing the old one first and then creating or setting another one. This brief window while there is no default causes PVCs without `StorageClassName` created at that time to not have any default, but due to the retroactive StorageClass assignment this way of changing defaults is safe.

Reviewer: ```suggestion
This behavior helps administrators change default storage class by removing the old one first and then creating or setting another one. This brief window while there is no default causes PVCs with an empty `storageClassName` created at that time to not have any default, but due to the retroactive StorageClass assignment this way of changing defaults is safe.
```

Author: Same as above, I intentionally used the word "without" because it means that the value is set to nil (meaning there's a key in yaml file but value is left out), or the key is not there.

That's why I mention `""` around L74 to explain it. I can only think of adding a sentence that would say that "without" means no key or no value to make it more clear.

Reviewer: We need to clarify this - I really didn't understand the sense of the existing document.

It helps a lot to use the active voice to explain what actions the control plane takes, and then follow that up with a higher-level explanation of what that means for the cluster operator.

Author: Ack. I've changed it be more clear on the high level and with each paragraph it gets more specific - up to the point where it explains the differences between no value, no key, and empty string value. 

@sftim PTAL and let me know if the current version is ok.

  more than one default is specified, the admission plugin forbids the creation of
  all PVCs.
* If the admission plugin is turned off, there is no notion of a default

>   StorageClass. All PVCs that have `storageClassName` set to `""` can be

Reviewer 2: Just to summarise the below two sentences and check if my understanding is right, do we mean to say that if the PVC has the `storageClassName`missing it can be later updated once the storage class becomes available but cannot be bound to a PV with the value `""`. Whereas if it's initially bound to `""` then it can only be bound to a PV with value `""` and if a storage class becomes available later we can't update the value?

  bound only to PVs that have `storageClassName` also set to `""`.

>   However, PVCs with missing `storageClassName` can be updated later once
>   default StorageClass becomes available. If the PVC gets updated it will no

Reviewer 3: ```suggestion
  However, PVCs with missing `storageClassName` can be updated later once a
  default StorageClass becomes available. If the PVC gets updated it will no
```

  longer bind to PVs that have `storageClassName` also set to `""`.

> See [Retroactive StorageClass assignment](#retroactive-storageclass-assignment) for more details.

Reviewer 4: In case `Retroactive StorageClass assignment` is not a feature gate, I prefer the following way.

```suggestion
See [retroactive StorageClass assignment](#retroactive-storageclass-assignment) for more details.
```

> See [retroactive StorageClass assignment](#retroactive-storageclass-assignment) for more details.

Reviewer 4: ```suggestion
See [retroactive default StorageClass assignment](#retroactive-default-storageclass-assignment) for more details.
```
The anchor needs to be modified too according to the changed head title :)

Also, how about changing all `retroactive StorageClass assignment` to `retroactive default StorageClass assignment` which is the exact name of KEP ? :)

Author: Sure, no hard opinion on the naming :)

  more than one default is specified, the admission plugin forbids the creation of
  all PVCs.
* If the admission plugin is turned off, there is no notion of a default
  StorageClass. All PVCs that have `storageClassName` set to `""` can be
  bound only to PVs that have `storageClassName` also set to `""`.
  However, PVCs with missing `storageClassName` can be updated later once
  default StorageClass becomes available. If the PVC gets updated it will no
  longer bind to PVs that have `storageClassName` also set to `""`.

>   See `RetroactiveDefaultStorageClass` [feature](/docs/concepts/storage/persistent-volumes/#retroactive-storageclass-assignment) for more details.

Reviewer: ```suggestion

See [Retroactive StorageClass assignment](#retroactive-storageclass-assignment) for more details.
```

Author: When writing this I've checked which form should be used and I found both where the feature/component name is **not** part of the link:

https://github.com/kubernetes/website/blob/13a3118ca5d206b95c295d3f2dcd4aa1a11778bb/content/en/docs/concepts/storage/persistent-volumes.md?plain=1#L58

And where it **is** part of the link:

https://github.com/kubernetes/website/blob/13a3118ca5d206b95c295d3f2dcd4aa1a11778bb/content/en/docs/concepts/storage/persistent-volumes.md?plain=1#L709

So I don't know which one is better. More opinions are welcome :)

Reviewer: I recommend writing it my way (I'm a tech lead for SIG Docs). In this case, the name of the feature gate is an unimportant detail.

of `storageClassName` attribute. This annotation is still working; however,
it won't be supported in a future Kubernetes release.

> #### Retroactive StorageClass assignment

Reviewer 4: I think adding `default` to this head is beneficial.
```suggestion
#### Retroactive default StorageClass assignment
```

However, if you enable the [`RetroactiveDefaultStorageClass` feature gate](/docs/reference/command-line-tools-reference/feature-gates/) then Kubernetes behaves differently: existing PVCs without `storageClassName` update to use the new default StorageClass.

When a default StorageClass becomes available, the control plane identifies any existing PVCs without `storageClassName`. For the PVCs that either have an empty value for `storageClassName` or do not have this key, the control plane then updates those PVCs to set `storageClassName` to match the new default StorageClass. If you have an existing PVC where the `storageClassName` is `""`, and you configure a default StorageClass, then this PVC will not get updated.

In order to keep binding to PVs with `storageClassName` set to `""` (while a default StorageClass is present), you need to set the `storageClassName` of the associated PVC to `""`.

> This behavior helps administrators change default storage class by removing the old one first and then creating or setting another one. This brief window while there is no default causes PVCs without `StorageClassName` created at that time to not have any default, but due to the retroactive StorageClass assignment this way of changing defaults is safe.

Reviewer 4: ```suggestion
This behavior helps administrators change default StorageClass by removing the old one first and then creating or setting another one. This brief window while there is no default causes PVCs without `storageClassName` created at that time to not have any default, but due to the retroactive StorageClass assignment this way of changing defaults is safe.
```
For consistency :)

> This behavior helps administrators change default StorageClass by removing the old one first and then creating or setting another one. This brief window while there is no default causes PVCs without `storageClassName` created at that time to not have any default, but due to the retroactive StorageClass assignment this way of changing defaults is safe.

Reviewer 4: ```suggestion
This behavior helps administrators change default StorageClass by removing the old one first and then creating or setting another one. This brief window while there is no default causes PVCs without `storageClassName` created at that time to not have any default, but due to the retroactive default StorageClass assignment this way of changing defaults is safe.
```
